JDK-8260332: ParallelGC: Cooperative pretouch for oldgen expansion

amitdpawar opened the pull request:
In case of ParallelGC, oldgen expansion can happen during promotion. Expanding thread will touch the pages and can't request for task execution as this GC thread is already executing a task. The expanding thread holds the lock on "ExpandHeap_lock" to resize the oldgen and other threads may wait for their turn. This is a blocking call.

This patch changes this behavior by adding another constructor in "MutexLocker" class to enable non blocking or try_lock operation. This way one thread will acquire the lock and other threads can join pretouch work. Threads failed to acquire the lock will join pretouch only when task is marked ready by expanding thread.

Following minimum expansion size are seen during expansion.
1. 512KB without largepages and without UseNUMA.
2. 64MB without largepages and with UseNUMA,
3. 2MB (on x86)  with large pages and without UseNUMA,
4. 64MB without large pages and with UseNUMA.

When Oldgen is expanding repeatedly with smaller size then this change wont help. For such cases, resize size should adapt to application demand to make use of this change. For example if application nature triggers 100 expansion with smaller sizes in same GC then it is better to increase the expansion size during each resize to reduce the number of resizes. If this patch is accepted then will plan to fix this case in another patch.

Jtreg all test passed.

Please review this change.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8260332](https://bugs.openjdk.java.net/browse/JDK-8260332): ParallelGC: Cooperative pretouch for oldgen expansion


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2976/head:pull/2976` \
`$ git checkout pull/2976`

Update a local copy of the PR: \
`$ git checkout pull/2976` \
`$ git pull https://git.openjdk.java.net/jdk pull/2976/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2976`

View PR using the GUI difftool: \
`$ git pr show -t 2976`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2976.diff">https://git.openjdk.java.net/jdk/pull/2976.diff</a>

</details>

amitdpawar commented:
Thank you for your feedback. I have updated as per your suggestion and fixed build issues on some targets. Please check.

dholmes-ora commented on src/hotspot/share/gc/shared/pretouchTask.cpp:
```diff
@@ -28,59 +28,99 @@
 
 void PretouchTask::work(uint worker_id) {
+
+  // Following atomic loads are required to make other processor store
+  // visible to all threads from this points.
```
That is not what an Atomic::load does; they only protect against word-tearing (which is only a theoretical possibility on some platforms). If you want to guarantee those loads see the most recent store then the fence needs to come first.

amitdpawar replied:
Thank you David for your reply and will do as suggested.

dholmes-ora commented on src/hotspot/share/gc/shared/pretouchTask.cpp:
```diff
@@ -28,59 +28,99 @@
+  thread_num = Atomic::sub(&_n_threads, 1u);
+
+  if (thread_num == 0) {
+    Atomic::release_store(&_cur_addr, _end_addr);
```
The release is unnecessary given the Atomic::sub.

kimbarrett replied:
Besides what David said, I think this is also confusing.  I think it's to account for the claim (using fetch_and_add) unconditionally adding the chunk size to cur_addr, which could lead to overshoot.  I think it would be clearer to prevent the overshoot from occurring.

amitdpawar replied:
David, will update as suggested.

Kim, thanks for your feedback. Updating **_cur_addr** was not needed here but thought to keep the markers in clean state. Again, **_cur_addr** & **_end_addr** gets updated before the start of next pretouch resize so will remove this. If something breaks then will fix as suggested. I hope this should be OK.

dholmes-ora commented on src/hotspot/share/gc/shared/pretouchTask.hpp:
```diff
@@ -30,15 +30,34 @@
+
+  void reinitialize(char* start_address, char *end_addr);
+
+  void set_task_status(TaskStatus status) { Atomic::release_store(&_task_status, (size_t)status);  }
+  void set_task_done()                    { Atomic::release_store(&_task_status, (size_t)Done);    }
```
There is a convention that when a setter embodies a release_store that release is included in the name e.g. release_set_task_done().

amitdpawar replied:
OK, will change.

dholmes-ora commented on src/hotspot/share/gc/shared/pretouchTask.hpp:
```diff
@@ -30,15 +30,34 @@
+  void set_task_done()                    { Atomic::release_store(&_task_status, (size_t)Done);    }
+  void set_task_ready()                   { set_task_status(Ready);    }
+  void set_task_notready()                { set_task_status(NotReady); }
+  bool is_task_ready()                    { return Atomic::load(&_task_status) == Ready; }
+  bool is_task_done()                     { return Atomic::load(&_task_status) == Done;  }
```
Given these fields are set with a release_store I would expect to see them read with a load_acquire to match with it.

amitdpawar replied:
OK, will change.

dholmes-ora commented on src/hotspot/share/gc/shared/pretouchTask.cpp:
```diff
@@ -28,59 +28,99 @@
+
+  if (thread_num == 0) {
+    Atomic::release_store(&_cur_addr, _end_addr);
+    OrderAccess::storestore();
+    set_task_done();
```
The storestore barrier is not needed as the set_task_done() is a release_store which already has a storestore barrier.

amitdpawar replied:
OK.

dholmes-ora commented on src/hotspot/share/gc/shared/pretouchTask.cpp:
```diff
@@ -28,59 +28,99 @@
 }
 
+void PretouchTask::reinitialize(char* start_addr, char* end_addr) {
+  Atomic::release_store(&_cur_addr, start_addr);
+  Atomic::release_store(&_end_addr, end_addr);
```
Back-to-back release-stores have redundant loadstore semantics. You could just use a storestore() barrier after the first release_store, then use plain Atomic::store for the second field.

amitdpawar replied:
OK, will change as per your suggestion.

dholmes-ora commented on src/hotspot/share/gc/parallel/mutableSpace.cpp:
```diff
@@ -121,11 +123,44 @@ void MutableSpace::initialize(MemRegion mr,
+        log_debug(gc, heap)("Running %s with " SIZE_FORMAT " work units pre-touching " SIZE_FORMAT "B.",
+                            pretouch_task->name(), num_chunks, total_bytes);
+
+        OrderAccess::storestore();
+        // Mark Pretouch task ready here and continue. The other threads
+        // waiting to expand old-gen will join from PSOldGen::expand_for_allocate
+        // function for pretouch work.
+        pretouch_task->set_task_ready();
```
The storestore is not needed as it is included in the release_store of set_task_ready().

amitdpawar replied:
OK.

amitdpawar commented on src/hotspot/share/gc/parallel/psOldGen.cpp:
```diff
@@ -182,17 +194,45 @@ bool PSOldGen::expand_for_allocate(size_t word_size) {
+        // Lets help expanding thread to pretouch the memory.
+        while (!pretouch()->is_task_done()) {
+          if (pretouch()->is_task_ready()) {
+            pretouch()->work(Thread::current()->osthread()->thread_id());
```
Is following change OK ?
pretouch()->work(static_cast<AbstractGangWorker*>(Thread::current())->id());

otherwise please suggest.

amitdpawar commented on src/hotspot/share/gc/shared/pretouchTask.cpp:
```diff
@@ -28,59 +28,99 @@
+  // Following atomic loads are required to make other processor store
+  // visible to all threads from this points.
+  char *cur_addr = Atomic::load(&_cur_addr);
+  char *end_addr = Atomic::load(&_end_addr);
```
didn't realize this. will change per your suggestion.

amitdpawar commented on src/hotspot/share/gc/parallel/psOldGen.hpp:
```diff
@@ -44,6 +46,7 @@ class PSOldGen : public CHeapObj<mtGC> {
   // Performance Counters
   PSGenerationCounters*    _gen_counters;
   SpaceCounters*           _space_counters;
+  PretouchTask*            _pretouch;          // Used when old gen resized during scavenging.
```
Thanks for your suggestion. It will be better if you can give some direction as G1GC also need similar fix during the expansion. I think G1 has array of regions to be touched during the expansion and that task needs to be shared across the threads. pretouch class can be put inside another class and that can be shared across the threads to synchronize the pretouch task and not sure whether that will be OK again. I explored this approach after your suggestion and thought to update after the feedback during the review process. Please suggest.

amitdpawar commented:
> 
> 
> _Mailing list message from [Kim Barrett](mailto:kim.barrett@oracle.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> > On Mar 12, 2021, at 3:01 PM, Amit Pawar <github.com+71302734+amitdpawar at openjdk.java.net> wrote:
> > In case of ParallelGC, oldgen expansion can happen during promotion. Expanding thread will touch the pages and can't request for task execution as this GC thread is already executing a task. The expanding thread holds the lock on "ExpandHeap_lock" to resize the oldgen and other threads may wait for their turn. This is a blocking call.
> > This patch changes this behavior by adding another constructor in "MutexLocker" class to enable non blocking or try_lock operation. This way one thread will acquire the lock and other threads can join pretouch work. Threads failed to acquire the lock will join pretouch only when task is marked ready by expanding thread.
> > Following minimum expansion size are seen during expansion.
> > 1. 512KB without largepages and without UseNUMA.
> > 2. 64MB without largepages and with UseNUMA,
> > 3. 2MB (on x86)  with large pages and without UseNUMA,
> > 4. 64MB without large pages and with UseNUMA.
> > When Oldgen is expanding repeatedly with smaller size then this change wont help. For such cases, resize size should adapt to application demand to make use of this change. For example if application nature triggers 100 expansion with smaller sizes in same GC then it is better to increase the expansion size during each resize to reduce the number of resizes. If this patch is accepted then will plan to fix this case in another patch.
> 
> Sorry, but a change like this needs better motivation. What you say
> above suggests this change doesn't actually help.
> 
> It's intentional that oldgen expansions aren't generally large, as the
> oldgen shouldn't be grown unnecessarily. There are already parameters
> such as MinHeapDeltaBytes to control and manipulate this.
> 
> It is also preferable to complete an expansion request quickly to make
> the additional space available to other threads in the main allocation
> path, rather than making them go to the expand path. Making expansions
> larger could force more threads to take the slower expand path, which
> doesn't seem like a win even if they then help with the pretouch part
> of another thread's expansion. (And that also assumes UsePreTouch is
> even enabled.)
> 
> So the followup change that you say is needed to make this one
> profitable seems questionable.
> 
> The proposed change is also surprisingly large and intrusive for
> something that seems like it should be very localized.
> 
> > Jtreg all test passed.
> 
> A change like this needs a lot more testing than that, both functionally
> and performance.

[https://bugs.openjdk.java.net/browse/JDK-8254699](JDK-8254699) contains test results in XL file to show PreTouchParallelChunkSize was recently changed from 1GB to 4MB on Linux after testing various sizes. I have downloaded the same XL file and same is updated for Oldgen case during resize and it gives some rough idea about the improvement for this fix and follow up fix. Please check "PretouchOldgenDuringResize" sheet for "Co-operative Fix" and "Adaptive Resize Fix" columns.

[PreTouchParallelChunkSize_TestResults_UpdatedForOldGenCase.xlsx](https://github.com/openjdk/jdk/files/6147180/PreTouchParallelChunkSize_TestResults_UpdatedForOldGenCase.xlsx)

Running SPECJbb composite shows 30-40% reduction in GC pause time when old-gen expands upto ~1GB for UseNUMA case (I tested for minimum oldgen size to trigger the resize). Non UseNUMA case will show improvement only when resize expands more than minimum "PreTouchParallelChunkSize" size to let other thread participate in pretouch work. Two cases 1 & 3 (without UseNUMA and with/without hugpages) above didnt shows any improvement because of "PreTouchParallelChunkSize" limit and that is the reason why I suggested another fix. 

The "Adaptive Resize Fix" column in the sheet is for next suggested fix and may possibly help to improve further. For server JVM, expansion size of 512KB,  2MB (hugepages) and 64MB looks good for first resize but later needs some attention I think. JVM flag "MinHeapDeltaBytes" needs to be known by the user and need to set it upfront. I think this can be consider for first resize in every GC and later dynamically go for higher size like double the previous size to adopt to application nature. This way it may help to reduce the GC pause time during the expansion. I thought to share my observation and my understanding could be wrong. So please check and suggest. 

Again, thanks for your feedback.

kimbarrett commented:
> [https://bugs.openjdk.java.net/browse/JDK-8254699](JDK-8254699) contains test results in XL file to show PreTouchParallelChunkSize was recently changed from 1GB to 4MB on Linux after testing various sizes. I have downloaded the same XL file and same is updated for Oldgen case during resize and it gives some rough idea about the improvement for this fix and follow up fix. Please check "PretouchOldgenDuringResize" sheet for "Co-operative Fix" and "Adaptive Resize Fix" columns.
> 
> [PreTouchParallelChunkSize_TestResults_UpdatedForOldGenCase.xlsx](https://github.com/openjdk/jdk/files/6147180/PreTouchParallelChunkSize_TestResults_UpdatedForOldGenCase.xlsx)

I'm not sure how to interpret this. In particular, it says "Test done using November 27rd Openjdk build". That predates a number of recent changes in this area that seem relevant. Is that correct? Or is that stale and the data has been updated for a newer baseline.

Also, can you provide details about how this information is generated and collected?  I would like to reproduce the results and do some other experiments.

I think that, as proposed, the change is making already messy code still more messy, and some refactoring is needed. I don't know what the details of that might look like yet; I'm still exploring and poking at the code. I also need to look at the corresponding part of G1 that you mentioned.

> The "Adaptive Resize Fix" column in the sheet is for next suggested fix and may possibly help to improve further. For server JVM, expansion size of 512KB, 2MB (hugepages) and 64MB looks good for first resize but later needs some attention I think. JVM flag "MinHeapDeltaBytes" needs to be known by the user and need to set it upfront. I think this can be consider for first resize in every GC and later dynamically go for higher size like double the previous size to adopt to application nature. This way it may help to reduce the GC pause time during the expansion. I thought to share my observation and my understanding could be wrong. So please check and suggest.

I think something like this has a serious risk of growing the oldgen a lot more than needed, which may have serious downsides.

amitdpawar commented:
Thank you Kim for your reply and please see my inline comments.
> 
> 
> > [https://bugs.openjdk.java.net/browse/JDK-8254699](JDK-8254699) contains test results in XL file to show PreTouchParallelChunkSize was recently changed from 1GB to 4MB on Linux after testing various sizes. I have downloaded the same XL file and same is updated for Oldgen case during resize and it gives some rough idea about the improvement for this fix and follow up fix. Please check "PretouchOldgenDuringResize" sheet for "Co-operative Fix" and "Adaptive Resize Fix" columns.
> > [PreTouchParallelChunkSize_TestResults_UpdatedForOldGenCase.xlsx](https://github.com/openjdk/jdk/files/6147180/PreTouchParallelChunkSize_TestResults_UpdatedForOldGenCase.xlsx)
> 
> I'm not sure how to interpret this. In particular, it says "Test done using November 27rd Openjdk build". That predates a number of recent changes in this area that seem relevant. Is that correct? Or is that stale and the data has been updated for a newer baseline.

Sorry for the confusion. I didnt test again and pre-touch time taken with different chunk size per thread was already recorded in the spreadsheet and thought to use it for reference to reply to David feedback "A change like this needs a lot more testing than that, both functionally and performance.".

> 
> Also, can you provide details about how this information is generated and collected? I would like to reproduce the results and do some other experiments.

JDK-8254699 was fixed by me and test was done using SPECJbb composite with following command along with JVM flag "PreTouchParallelChunkSize" with values 1G (earlier default on x86) and 4M (current default on x86). To generate this info,  "Ticks::now()" function was called to record the time taken for pretouch operation. The first table (please ignore resize related) in the sheet "SPECjbb_Summary"  was created from the sheet "SPECjbb_128C256T_GCPretouchTime" that contains pretouch log dumps for SPECjbb composite tests. 

Command (also mentioned in SPECjbb_128C256T_GCPretouchTime sheet):
-Xms24g -Xmx970g -Xmn960g -server -XX:+UseParallelGC -XX:+AlwaysPreTouch -XX:+UseLargePages -XX:LargePageSizeInBytes=2m -XX:MaxTenuringThreshold=15 -XX:+ScavengeBeforeFullGC -XX:+UseAdaptiveSizePolicy -XX:ParallelGCThreads=256 -XX:-UseBiasedLocking -XX:SurvivorRatio=200 -XX:TargetSurvivorRatio=95 -XX:+UseNUMA -XX:-UseNUMAInterleaving -XX:+UseTransparentHugePages -XX:-UseAdaptiveNUMAChunkSizing

> 
> I think that, as proposed, the change is making already messy code still more messy, and some refactoring is needed. I don't know what the details of that might look like yet; I'm still exploring and poking at the code. I also need to look at the corresponding part of G1 that you mentioned.

First, I thought to push this change for ParallelGC and next G1GC. Earlier ParallelGC pretouch was done with single threaded and later this was fixed by moving multithreaded pretouch common code from G1GC and ParallelGC to PretouchTask class. I thought to do it similarly for this case too.

> 
> > The "Adaptive Resize Fix" column in the sheet is for next suggested fix and may possibly help to improve further. For server JVM, expansion size of 512KB, 2MB (hugepages) and 64MB looks good for first resize but later needs some attention I think. JVM flag "MinHeapDeltaBytes" needs to be known by the user and need to set it upfront. I think this can be consider for first resize in every GC and later dynamically go for higher size like double the previous size to adopt to application nature. This way it may help to reduce the GC pause time during the expansion. I thought to share my observation and my understanding could be wrong. So please check and suggest.
> 
> I think something like this has a serious risk of growing the oldgen a lot more than needed, which may have serious downsides.

I believe this depends on application runtime nature and it may expand as needed. If you still think that this change is not useful then please suggest. Will withdraw this PR.

Thanks,
Amit

kimbarrett commented:
> > > [PreTouchParallelChunkSize_TestResults_UpdatedForOldGenCase.xlsx](https://github.com/openjdk/jdk/files/6147180/PreTouchParallelChunkSize_TestResults_UpdatedForOldGenCase.xlsx)
> > 
> > 
> > I'm not sure how to interpret this. In particular, it says "Test done using November 27rd Openjdk build". That predates a number of recent changes in this area that seem relevant. Is that correct? Or is that stale and the data has been updated for a newer baseline.
> 
> Sorry for the confusion. I didnt test again and pre-touch time taken with different chunk size per thread was already recorded in the spreadsheet and thought to use it for reference to reply to David feedback "A change like this needs a lot more testing than that, both functionally and performance.".

Getting any benefit from parallel pretouch here suggests we have many threads piling up on the expand mutex. Before JDK-8260045 (pushed together with JDK-8260044 on 2/12/2021) that would lead to "expansion storms", where several threads piled up on that mutex and serially entered and did a new expansion (with associated pretouch, and possibly of significant size). Getting rid of the expansion storms may alleviate the serial pretouch quite a bit. There may still be some benefit to cooperative parallization, but new measurements are needed to determine how much of a problem still exists.

If it is still worth addressing, I think the proposed approach has problems, as it makes messy complicated code messier and more complicated.  I think some preliminary refactoring is needed.  For example, splitting MutableSpace::initialize into a setup-pages part and set-the-range part.  I've not explored in detail what's needed yet, pending new measurements.  (I haven't had time to do those measurements yet myself.)

amitdpawar commented:
Thanks Kim for your feedback.

I have updated the code and it looks good now. Please use "Full" diff to see all the changes.

Summary of changes: added new class in pretouchTask.hpp file for synchronization. Now co-operative pretouch is enabled for both ParallelGC and G1GC. For testing purpose, added new flag "UseMultithreadedPretouchForOldGen" and timing is enabled for old-gen pretouch case for measurements. Enabling this flag makes try_lock to be used instead of lock (which is current default).

This change will be useful only if PreTouchParallelChunkSize value is not very huge compared to old-gen expansion size. For Linux it is set to 4MB and for other OS it is still 1GB.

Testing SPECJbb composite on AMD EPYC machine shows following improvement.

- Parallel GC: with UseNUMA, minimum expansion size is 64MB
-- with 2MB hugepages, master branch time taken is ~9-10ms and current fix time taken is ~1-2ms`
-- without hugepages, master branch time taken is ~14-17ms and current fix time taken is ~2-3ms

- Parallel GC: without UseNUMA, minimum expansion size is 512KB so this change is not useful for this size. Sometime expansion goes upto 20MB then pretouch takes less time for such large size.

- G1 GC:  with UseNUMA, maximum expansion size is 32MB
-- with 2MB hugepages, master branch time taken is ~5-7ms and current fix time taken is ~1-7ms` (higher time due to single thread activity as there was no other thread waiting to acquire the lock)

- G1 GC:  without UseNUMA, minimum expansion size 64KB and maximum expansion size is 32MB
-- without hugepages, master branch time taken = ~9-11ms and current fix time taken ~2-9ms (same here too)

Please check.

amitdpawar commented:
Fixing runtime issue seen in Pre-submit tests. I somehow missed no safepoint check while acquiring the lock in G1 case. Minor issue. Thanks.

kimbarrett commented:
Sorry to be slow to respond.  I've done some prototyping of an alternative that I think is promising, but have been swamped with other work.  It might be a while before I can really spend time on it again.

amitdpawar commented:
Thanks for your reply and good to know that alternative implementation is possible. I will wait. Thanks
